lib/LedgerSMB/Upgrade_Tests.pm

```diff
@@ -415,6 +472,94 @@ database, or delete the offending rows through PgAdmin III or psql'),
   max_version => '1.4'
 );
 
+push @tests, __PACKAGE__->new(
+   test_query => 'select *, eca.id as id  from entity_credit_account eca
+                     join entity_class ec on eca.entity_class = ec.id
+                     join entity e on eca.entity_id = e.id
+                   where meta_number in
+                       (select meta_number from entity_credit_account
+                        group by meta_number having count(*) > 1)
+                   order by meta_number',
+ display_name => marktext('No duplicate meta_numbers'),
+         name => 'no_meta_number_dupes',
+ display_cols => [ 'meta_number', 'class', 'description', 'name' ],
+      columns => ['meta_number'],
+        table => 'entity_credit_account',
+ instructions => marktext('Make sure all meta numbers are unique.'),
+      appname => 'ledgersmb',
+  min_version => '1.3',
+  max_version => '1.4'
+);
+
+push @tests, __PACKAGE__->new(
+   test_query => q(select distinct gifi_accno from chart
+                   where not exists (select 1
+                                       from gifi
+                                      where gifi.accno = chart.gifi_accno)
+                         and gifi_accno is not null
+                         and gifi_accno !~ '^\\s*\$'),
+ display_name => marktext('GIFI accounts not in "gifi" table'),
+         name => 'missing_gifi_table_rows',
+ display_cols => [ 'gifi_accno' ],
+        table => 'chart',
+ instructions => marktext('Please use the 1.2 UI to add the GIFI accounts'),
+      appname => 'ledgersmb',
+  min_version => '1.2',
+  max_version => '1.2'
+);
+
+push @tests, __PACKAGE__->new(
+   test_query => q(select distinct gifi_accno from account
+                   where not exists (select 1
+                                       from gifi
+                                      where gifi.accno = account.gifi_accno)
+                         and gifi_accno is not null
+                         and gifi_accno !~ '^\\s*\$'),
+ display_name => marktext('GIFI accounts not in "gifi" table'),
+         name => 'missing_gifi_table_rows',
+ display_cols => [ 'gifi_accno' ],
+        table => 'account',
+ instructions => marktext('Please use the 1.3/1.4 UI to add the GIFI accounts'),
+      appname => 'ledgersmb',
+  min_version => '1.3',
+  max_version => '1.4'
+);
+
+#=pod
+
+# This must be done before the acc_trans test, for they have duplicates
+# and need a unique key
+    push @tests, __PACKAGE__->new(
+        # Add a unique key to allow editing
+        # Make the test serially reusable and protect against triggers
+        test_query => 'ALTER TABLE acc_trans DISABLE TRIGGER USER;
```

Comment:
While I understand the problem, I'm not too hot about this solution: it abuses the validation system to execute migration code. 

Can you indicate what happens when this code is run twice?

Reply:
Fixed, moved into pre-tests.